### ISSUE TITLE
Clippy ignores crate wide warn/deny

### ISSUE THREAD

**Issue by tcharding** [rust-lang/rust-clippy#6610]
I put this code in `main.rs`

```rust
#![deny(clippy::print_stdout)]

mod foo;

#[allow(clippy::print_stdout)]
fn main() {
    println!("Hello, world!");
}
```
And this code in `foo.rs`
```rust
pub fn foo() {
    println!("This should trigger clippy");
}
```
I expected to see clippy warn something like
```

error: use of `println!`
 --> src/foo.rs:2:5
  |
6 |     println!("This should trigger clippy");
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
```
Instead all I get is 
```
warning: function is never used: `foo`
  ```

### Meta
```
cargo clippy -V                                                                                                            ✘ 101 
clippy 0.1.51 (4253153 2021-01-17)
rustc -Vv
rustc 1.51.0-nightly (4253153db 2021-01-17)
binary: rustc
commit-hash: 4253153db205251f72ea4493687a31e04a2a8ca0
commit-date: 2021-01-17
host: x86_64-unknown-linux-gnu
release: 1.51.0-nightly
LLVM version: 11.0.1
  ```

**Comment by giraffate** [comment]
It seems to be similar to https://github.com/rust-lang/rust-clippy/issues/5356.

**Comment by flip1995** [comment]
Yes this is the same underlying issue. The solution is to get rid of `pre_expansion_pass` lints.

**Comment by davepacheco** [comment]
I think I ran into this too and reproduced it here: https://github.com/davepacheco/clippy-issue

@giraffate and @flip1995 you said this is the same underlying issue as #5356, but that one looks like it's been fixed.  Do you expect this issue is fixed too?  I'm still seeing it on clippy 0.0.212 (e1884a8e 2020-12-29) (Rust stable 1.49.0).

This doesn't always seem to be a problem and I'm not sure why not.  I have a different crate (dropshot) that uses a crate-wide `#![allow(clippy::style)]` that appears to work.

**Comment by flip1995** [comment]
It's the same underlying issue, but requires a separate fix, which involves quite a bit of work. TL;DR the whole lint has to be rewritten.

**Comment by davepacheco** [comment]
Thanks for clarifying!

**Comment by jszwedko** [comment]
In case it helps others, as a work around I ended up adding these lints to `.cargo/config.toml`:

```toml
[target.'cfg(feature = "cargo-clippy")']
rustflags = [
  "-Dclippy::print_stdout",
  "-Dclippy::print_stderr",
  "-Dclippy::dbg_macro",
]
```

**Comment by MrMuetze** [comment]
My team just stumbled over this problem and we were wondering if we could supply any more information to help solve this issue? Crate wide clippy configuration is quite important for us :)

**Comment by flip1995** [comment]
This PR https://github.com/rust-lang/rust-clippy/pull/8518 will reduce the amount of lints that are affected by this issue to this set of lints:

    DEPRECATED_CFG_ATTR,
    MISMATCHED_TARGET_OS,
    EMPTY_LINE_AFTER_OUTER_ATTR,

Are you affected by one of those lints? If not #8518 should fix your issue.

We know why this happens, fixing this is just a huge amount of work or partly impossible, unless we accept major tradeoffs in lint quality. #8518 rewrites multiple lints completely, to give a approximate amount of work that is involved in fixing this.

**Comment by MrMuetze** [comment]
Oh perfect, thank you for the quick heads up. I was not aware of PR #8518 and how it would affect this issue. We do not use the lints you mentioned, so we should be golden as soon as that is merged :)

**Comment by tcharding** [comment]
Thanks for your efforts on this and also the explanations. Can be closed or left open as you think best suits the project.

**Comment by dsherret** [comment]
This seems to be working now in libraries at least where it wasn't before. I've started using it and it's great.

**Comment by jdonszelmann** [comment]
fixed, just like #7549

**Comment by jdonszelmann** [comment]
oops, this was noted earlier and kept open for visibility, which makes sense. I marked some others (near-duplicates) as fixed too. I am going through the issues since I think I fixed the root cause and trying to find issues to reproduce the old behavior.
